### DISCUSSION TITLE
Why `Button.Click` not call

### DISCUSSION THREAD

**Discussion by Coloryr** [AvaloniaUI/Avalonia#10165] [Q&A]
Version: `11.0.999-cibuild0029429-beta`
but work in `11.0.0-preview4`

```
    public MainWindow()
    {
        InitializeComponent();

        Head.SetWindow(this);

```

```
    private Window BindWindow;

    public void SetWindow(Window window)
    {
        BindWindow = window;
    }

    public HeadControl()
    {
        InitializeComponent();

        DataContext = this;

        PointerPressed += HeadControl_PointerPressed;

        Button_Min.Click += ButtonMin_Click;
        Button_Max.Click += ButtonMax_Click;
        Button_Close.Click += ButtonClose_Click;
    }
```

When `BindWindow = window;`, the `Button.Click` are not call.

**Comment by Coloryr** [comment]
![2YRPT6E3`T@~}W790 587M4](https://user-images.githubusercontent.com/26276037/216280754-84090779-5420-431b-abf6-e6ff9bb70ad2.png)
![`3Z6EGAPZR5X$3EEEQW(L@J](https://user-images.githubusercontent.com/26276037/216280852-254bda0c-70f8-4574-a77f-3f8305ea8ed8.png)

**Comment by Coloryr** [comment]
![IIF13JMT98 3Q`@IMOMXR1G](https://user-images.githubusercontent.com/26276037/216281195-1ca2f01a-4a12-4d33-902f-dfdf0743876c.png)

**Comment by Coloryr** [comment]
```
<Window
    x:Class="ColorMC.Gui.UI.Windows.MainWindow"
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:setting="clr-namespace:ColorMC.Gui.Utils.LaunchSetting"
    xmlns:views="clr-namespace:ColorMC.Gui.UI.Controls"
    xmlns:views1="clr-namespace:ColorMC.Gui.UI.Controls.Main"
    Title="ColorMC"
    Width="800"
    Height="450"
    MinWidth="800"
    MinHeight="450"
    d:DesignHeight="450"
    d:DesignWidth="800"
    Background="Transparent"
    ExtendClientAreaChromeHints="NoChrome"
    ExtendClientAreaTitleBarHeightHint="-1"
    ExtendClientAreaToDecorationsHint="True"
    SystemDecorations="BorderOnly"
    mc:Ignorable="d">
    <Grid>
        <views:HeadControl
            Name="Head"
            Title="{setting:Localize MainWindow.Title}"
            HorizontalAlignment="Stretch"
            VerticalAlignment="Top" />
        <Grid Name="Grid1" Margin="0,35,0,0">
```

```
<UserControl
    x:Class="ColorMC.Gui.UI.Controls.HeadControl"
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:setting="clr-namespace:ColorMC.Gui.Utils.LaunchSetting"
    Height="35"
    d:DesignHeight="35"
    d:DesignWidth="800"
    Background="{setting:Colors Main}"
    mc:Ignorable="d">
    <Grid Margin="5,5,5,0">
```

**Reply by timunie** [reply]
If HeadControl is meant to be a re-usable control, you can better use `ContentControl` or `HeaderedContentControl` as your base-class. 

`UserControl` can be seen as a page of content.

**Reply by Coloryr** [reply]
> If HeadControl is meant to be a re-usable control, you can better use `ContentControl` or `HeaderedContentControl` as your base-class.
> 
> `UserControl` can be seen as a page of content.

Switch to ` ContentControl` not work

![image](https://user-images.githubusercontent.com/26276037/216286141-025cda99-3a9c-479b-a2be-84ac166f5aec.png)

**Reply by timunie** [reply]
https://github.com/timunie/Avalonia.Samples/tree/Controls/RatingControl/src/Avalonia.Samples/CustomControls/RatingControlSample 

check out this tutorial. Note: it's WIP and not finished yet, but may give you a starting point.

**Reply by Coloryr** [reply]
> https://github.com/timunie/Avalonia.Samples/tree/Controls/RatingControl/src/Avalonia.Samples/CustomControls/RatingControlSample
> 
> check out this tutorial. Note: it's WIP and not finished yet, but may give you a starting point.

I test the old Pack, the event can be call in `11.0.999-cibuild0029192-beta`.

**Reply by Coloryr** [reply]
No work in `11.0.999-cibuild0029361-beta`

**Reply by timunie** [reply]
Do you have a minimal App to reproduce?

**Reply by Coloryr** [reply]
Only juet set a field than the event not work...

**Reply by Coloryr** [reply]
> Do you have a minimal App to reproduce?

https://github.com/Coloryr/ColorMC/tree/what/src

**Reply by timunie** [reply]
That's a whole App with many many code inside. I have neither the time to debug this entirely nor am I allowed to run this code without checking it for safety first. So to help, I really need a minimal sample only showing the not working button. 

Thank you for your understanding. 

But what I can see, you often tend to call `Wait()` and `.Result` for async operations. This can be  the root cause of many issues as The process may never finish. Try to read some tutorials and docs about `async await` in general.

**Reply by Coloryr** [reply]
> That's a whole App with many many code inside. I have neither the time to debug this entirely nor am I allowed to run this code without checking it for safety first. So to help, I really need a minimal sample only showing the not working button.
> 
> Thank you for your understanding.
> 
> But what I can see, you often tend to call `Wait()` and `.Result` for async operations. This can be the root cause of many issues as The process may never finish. Try to read some tutorials and docs about `async await` in general.

I try to write a `not working button` project.

**Reply by Coloryr** [reply]
[AvaloniaApplication1.zip](https://github.com/AvaloniaUI/Avalonia/files/10569900/AvaloniaApplication1.zip)
The code.
And 
![(}L}6_2Y5CU5RY}@ZZ(MSY5](https://user-images.githubusercontent.com/26276037/216350716-dfe608a3-77b0-4d77-bf50-d495cd03499b.png)
There button no work if

![P%C}G$5IXEW`NN6R(QUF5LB](https://user-images.githubusercontent.com/26276037/216350846-b23fe619-96eb-47e2-95f4-862c79a0dc1f.png)

**Reply by Coloryr** [reply]
![L`%X1X CN`Z$@)%YY G)L8](https://user-images.githubusercontent.com/26276037/216351095-48b63ec8-b42d-4483-8b5c-aff29b11dc93.png)

**Reply by Coloryr** [reply]
Whatever, very thankful for you help.

**Reply by timunie** [reply]
Okay, It's an easy fix. So what happens is, that your entire header can be used to move the window around. This includes the buttons. If the user clicks the button, draggin will be started and the event is marked as handled. The Button will never recieve the click. So I did the following to solve:

First added another control above the title column:

```diff
-    <Grid Margin="5,5,5,0" >
+    <Grid Margin="5,5,5,0" ColumnDefinitions="*, Auto">
        <Label
            Name="TitleShow"
            Margin="10,0,0,0"
            Content="{Binding Title}"
            Foreground="White" />
+        <!-- Gripper control, Transparent in order to be Hittest-visible -->
+        <Panel x:Name="Gripper"
+               Background="Transparent"
+               Grid.Column="0" />
        <StackPanel
+            Grid.Column="1"
            HorizontalAlignment="Right"
            VerticalAlignment="Top"
            Orientation="Horizontal">
            <Button
                Name="Button_Min"
                Width="24"
                Height="24"
                Margin="0,0,5,0"
                Content="-" />
            <Button

[...] 

```

Then I use the gripper to handle the drag, not the entire header:

```diff
    public HeadControl()
    {
        InitializeComponent();

        DataContext = this;

+        Gripper.PointerPressed += Gripper_PointerPressed;
        Button_Min.Click += ButtonMin_Click;
        Button_Max.Click += ButtonMax_Click;
        Button_Close.Click += ButtonClose_Click;
    }

+    private void Gripper_PointerPressed(object? sender, PointerPressedEventArgs e)
+    {
+        if (BindWindow == null)
+            return;
+
+        BindWindow.BeginMoveDrag(e);
+    }
```

I hope that makes sense.

**Reply by Coloryr** [reply]
Thanks
